### ISSUE TITLE
Build and Layout investigation

### ISSUE THREAD

**Issue by dlabaj** [patternfly/patternfly-infra-issues#120]
## Description
As a developer we would like to know how the Svelte code is generated to HTML, and if it will fit our documentation needs.

### To Do
- [x] Show how the output of the Svelte code looks in HTML. Does the build process give us minified code? Is there a way to get non minified code from vite and Svelte? 
- [x] Decide how we can use the output in our documentation framework. Are we going to generate all HTML and consume that? Is it going to be generated into markdown?
- [x] Document pros and cons of moving to Svelte so the team can have a way to evaluate it compared to alternative solutions.
- [x] Flush out the infrastructure and how a developer would add a component in core using Svelte.

### Resources
[Svelte Documentation](https://svelte.dev/docs/svelte/overview)
[Vite Documentation](https://vite.dev/guide/)

**Comment by mattnolting** [comment]
import Handlebars from 'handlebars';
import fs from 'fs';
import path from 'path';

class AlertTemplateManager {
  constructor() {
    this.hbs = Handlebars.create();
    this.templates = {};
    this.partials = {};
  }

  // Load base template
  loadBaseTemplate(templatePath) {
    const templateContent = fs.readFileSync(templatePath, 'utf8');
    this.templates.base = this.hbs.compile(templateContent);
  }

  // Register partials
  registerPartials(partialsDir) {
    const partialFiles = fs.readdirSync(partialsDir)
      .filter(file => file.endsWith('.hbs'));

    partialFiles.forEach(file => {
      const partialName = path.basename(file, '.hbs');
      const partialPath = path.join(partialsDir, file);
      const partialContent = fs.readFileSync(partialPath, 'utf8');

      this.hbs.registerPartial(partialName, partialContent);
      this.partials[partialName] = partialContent;
    });
  }

  // Render alert with context
  renderHbs(context) {
    // Validate required fields
    if (!context.title) {
      throw new Error(`Component title is required`);
    }

    // Merge with default context
    const defaultContext = {
      description: null,
      actionGroup: null,
      modifier: 'pf-m-success',
      iconType: 'check-circle'
    };

    const mergedContext = { ...defaultContext, ...context };

    // Render the template
    return this.templates.base(mergedContext);
  }

  // Create preset alerts
  createPresetAlerts() {
    return {
      success: this.renderHbs({
        title: 'Operation Successful',
        description: 'Your action was completed successfully.',
        modifier: 'pf-m-success'
      }),
      error: this.renderHbs({
        title: 'Error Occurred',
        description: 'There was a problem with your request.',
        modifier: 'pf-m-danger',
        iconType: 'exclamation-circle'
      }),
      warning: this.renderHbs({
        title: 'Warning',
        description: 'Please proceed with caution.',
        modifier: 'pf-m-warning',
        iconType: 'exclamation-triangle'
      })
    };
  }
}

// Export as singleton
export default new AlertTemplateManager();

**Comment by mattnolting** [comment]
# Template System Migration Guide: From Handlebars to SvelteKit

## Executive Summary
Think of this like replacing a monolithic templating engine with microservices - while the system is still running. We're migrating from Handlebars to SvelteKit, but instead of a risky big-bang migration, we're taking the strangler fig pattern approach. The old system keeps running while we gradually wrap new functionality around it.

**Core Strategy**
- Keep the template manager as our abstraction layer
- Migrate components individually, like deploying microservices
- No downtime, no breaking changes
- Incremental improvements, testable at each step

## The Migration Pattern

Similar to how Git allows us to maintain multiple branches while developing new features, our migration maintains dual implementations during transition:

```
main (production) ----*----*----*----> (Handlebars)
                      \    \    \
feature branches       *----*----* --> (SvelteKit)
```

## Architecture Evolution

### Current: Handlebars System
```
templates/
└── base templates
    └── partial templates
        └── component templates
```

### Target: SvelteKit Architecture
```
src/
├── lib/modules/
│   ├── components/      # Component library
│   ├── hbs/            # Legacy support
│   └── layouts/        # Layout system
└── routes/             # Page routing
```

## Migration Path

### Phase 1: Foundation
```
.svelte-kit/
└── src/
    └── lib/
        ├── modules/
        │   ├── components/    # New component system
        │   ├── hbs/          # Legacy template support
        │   ├── layouts/      # Layout architecture
        │   └── utilities/    # Shared utilities
        └── Templates/        # Base templates
```

### Phase 2: Component Migration
```javascript
// Before: Handlebars Template
{{!-- alert-template.hbs --}}
<div class="alert {{modifier}}">
  {{content}}
</div>

// After: Svelte Component
<!-- Alert.svelte -->
<script>
  export let modifier = '';
  export let content = '';
</script>

<div class="alert {modifier}">
  {content}
</div>
```

## Implementation Strategy

### Template Manager Abstraction
```javascript
class ModernTemplateManager {
  constructor() {
    this.templates = new Map();
    this.validators = new Map();
  }

  // Supports both systems during migration
  async render(name, context) {
    const template = this.templates.get(name);
    if (template.type === 'hbs') {
      return this.renderHandlebars(template, context);
    }
    return this.renderSvelte(template, context);
  }
}
```

## Migration Workflow

1. Package Structure
   ```
   src/lib/modules/components/
   ├── Alert/
   │   ├── Alert.svelte    # New implementation
   │   └── alert.hbs      # Legacy support
   ```

2. Route Structure
   ```
   src/routes/
   ├── docs/
   │   └── components/
   │       └── alert/
   │           ├── +layout.svelte
   │           └── +page.svelte
   ```

3. Documentation Integration
   ```svelte
   <!-- +page.svelte -->
   <script>
     import { Alert } from '$lib/modules/components';
   </script>

   # Alert Component
   <Alert type="success">
     Migration successful!
   </Alert>
   ```

## Success Metrics

- Zero downtime during migration
- Component-by-component transition
- Maintained documentation coverage
- Improved developer experience

## Rollback Strategy

Components maintain dual implementations until fully migrated:
```javascript
render(name, context) {
  return this.useNewSystem
    ? this.renderSvelte(name, context)
    : this.renderHandlebars(name, context);
}
```

## Developer Guide

1. Setup New Component
   ```bash
   └── src/
       └── lib/
           └── modules/
               └── YOUR_COMPONENT/
                   ├── index.ts
                   ├── Component.svelte
                   └── legacy.hbs
   ```

2. Migration Steps
   ```javascript
   // 1. Create Svelte version
   // 2. Test in isolation
   // 3. Update documentation
   // 4. Remove legacy template
   ```

**Comment by mattnolting** [comment]
# Migration Effort Analysis

## Compatibility Pattern
```
From Handlebars To:
--------------------
Svelte:  [===================] 95% automatable
JSX:     [=======------------] 70% manual conversion
Astro:   [========-----------] 80% partial reuse
```

## Migration Path Effort

```
Svelte Approach:
--------------------
1. Automated conversion     → 60% of templates
2. Parallel systems        → Zero downtime
3. Component-by-component  → Gradual rollout
4. Feature toggles         → Instant rollback
Total effort: 40% of full rewrite

JSX Approach:
--------------------
1. Full rewrite required   → 100% of templates
2. Big bang deployment     → Potential downtime
3. System-wide changes     → All or nothing
4. Complex state transfer  → High risk
Total effort: 100% manual conversion

Astro Approach:
--------------------
1. Partial template reuse  → 50% of templates
2. Mixed deployment        → Some downtime
3. Page-by-page migration  → Staged rollout
4. Limited rollback        → Medium risk
Total effort: 70% of full rewrite
```

## Automation Potential

```
Handlebars → Svelte Script:
--------------------
{{#if condition}}     →  {#if condition}
{{content}}           →  {content}
{{/if}}               →  {/if}

<div class="{{cls}}"> →  <div class={cls}>
{{#each items}}       →  {#each items as item}
{{/each}}             →  {/each}
```

## Migration Script Concept
```
Migration Tool:
--------------------
1. Parse Handlebars templates
2. Convert syntax patterns
3. Generate Svelte components
4. Maintain existing logic
5. Create parallel structures
6. Enable feature toggles

Remaining Manual Work:
--------------------
1. Complex logic conversion
2. State management updates
3. Event handler adjustments
4. Component optimization
```

This approach significantly reduces migration effort because:
- Most templates can be automatically converted
- No system downtime required
- Risk is minimized through parallel running
- Teams can validate each component individually
- Rollback is always available
- Business continues uninterrupted

**Comment by mattnolting** [comment]
# The Business Case for Migration

## Why Change is Necessary

Our current Handlebars system, while functional, is costing us significant time and money through hidden inefficiencies. Let's talk about what this means day-to-day:

### Time is Money
Developers currently spend more time wrestling with the system than building features. When a bug appears, they're forced to play detective across multiple repositories and files just to find where the issue originates. This isn't just frustrating - it's expensive. Every hour spent searching through templates and partials is an hour not spent delivering value.

### The Maintenance Nightmare
Our maintenance situation is precarious. Templates are scattered across repositories with no clear ownership. Documentation is sparse and often outdated. This creates several critical business risks:
- Knowledge is locked in the heads of a few developers
- Onboarding new team members takes weeks longer than it should
- Making system-wide changes is risky and time-consuming
- Bug fixes often introduce new problems due to unclear dependencies

### Developer Experience Matters
The current framework actively works against our developers. Simple tasks become complex operations. Modern development practices are difficult or impossible to implement. This leads to:
- Increased frustration and potential turnover
- Longer development cycles
- Higher error rates
- Resistance to making necessary changes

### Future-Proofing Our Investment
Handlebars is essentially frozen in time. While the rest of the web development world has moved forward with modern practices and tooling, we're stuck with:
- Limited extensibility
- No modern development tooling
- Poor testing capabilities
- Outdated development practices

## The Svelte Advantage

By moving to Svelte, we're not just updating our technology - we're investing in our development team's productivity and our product's future. Here's what changes:

### Immediate Benefits
- Developers can focus on building features instead of fighting the system
- Changes are isolated and testable, reducing risk
- Modern tooling enables faster development and better quality
- Clear documentation and active community support

### Long-term Gains
- Easier recruitment - developers want to work with modern technology
- Better retention - less frustration means happier developers
- Faster feature delivery - less system overhead means more time for actual development
- Reduced risks - better testing and isolation means fewer critical bugs

## The bottom line? 
This migration isn't just a technical upgrade - it's a business investment that will pay dividends in faster development, better quality, and improved team satisfaction.

Would you like me to elaborate on any of these points or discuss specific examples of how these benefits translate to business value?

**Comment by mattnolting** [comment]
# Vite + Svelte: Development Experience

## Hot Module Replacement (HMR)

### Current Process
When a developer makes a change:
```
Change file → Full page reload → Lose application state → Navigate back to spot
                  ↓
        Developer breaks focus
                  ↓
     Loses minutes per code change
```

### With Vite + Svelte
When a developer makes a change:
```
Change file → Instant update → State preserved → Continue working
                  ↓
      Developer stays focused
                  ↓
    Continuous flow maintained
```

## Build Performance

### Development Build Times
```
Traditional Build:
Initial Build:    45-120 seconds
Rebuild:          10-30 seconds
State:            Lost on every change

Vite:
Initial Start:    < 1 second
Updates:          < 50ms
State:            Preserved
```

### Production Build Times
```
Traditional Build:
Full Build:       3-5 minutes
Code Splitting:   Manual configuration
Cache Handling:   Complex setup

Vite:
Full Build:       30-60 seconds
Code Splitting:   Automatic
Cache Handling:   Built-in optimization
```

The impact? A developer making 50 changes per day saves:
- 25+ minutes in build waiting time
- 30+ minutes in lost context switching
- Hours in reduced frustration and maintained focus

This isn't just about faster builds - it's about maintaining developer flow state and eliminating the micro-frustrations that kill productivity.

**Comment by srambach** [comment]
I'm curious about automatically converting templates, both how that works and how you figured out the number that can be converted automatically.

Another thing I was wondering about was our handlebars helper functions - are there equivalents or is that something we'd also transition over, or code around?

**Comment by dlabaj** [comment]
@mattnolting Thanks for putting this together there's a lot of good details here. Do you mind putting up a PR with a **svelte-handlebars-spike.md** and moving the documentation to this [docs/spikes folder](https://github.com/patternfly/patternfly-infra-issues/tree/main/docs/spikes) I setup in this project? This way we all can leave feedback, have a discussion, and review it. Thanks again.
